client/pages/add-exercises.jsx

```diff
@@ -5,33 +5,108 @@ export default class AddExercisePage extends React.Component {
     super(props);
     this.state = {
       exerciseName: '',
-      exercises: []
+      exercises: [],
+      data: {
+        exercises: []
+      }
     };
     this.handleExerciseNameChange = this.handleExerciseNameChange.bind(this);
     this.handleSubmit = this.handleSubmit.bind(this);
     this.handleClick = this.handleClick.bind(this);
+    this.handleRepsChange = this.handleRepsChange.bind(this);
+    this.handleWeightChange = this.handleWeightChange.bind(this);
+    this.handleIconChange = this.handleIconChange.bind(this);
   }
 
   handleExerciseNameChange(event) {
     this.setState({ exerciseName: event.target.value });
   }
 
+  handleRepsChange(index, event) {
+    const dataCopy = Object.assign({}, this.state.data);
+    dataCopy.exercises[index].reps = event.target.value;
+    this.setState({ data: dataCopy });
+  }
+
+  handleWeightChange(index, event) {
+    const dataCopy = Object.assign({}, this.state.data);
+    dataCopy.exercises[index].weight = event.target.value;
+    this.setState({ data: dataCopy });
+  }
```

Comment:
we have to be more surgical when updating nested state values in react apps. we can't just copy the root object and then go buck wild on its contents. every level of nesting has to be copied.

```js
const dataCopy = Object.assigne({}, this.state.data);
dataCopy.exercises = dataCopy.exercises.slice();
dataCopy.exercises[index] = Object.assign({}, dataCopy.exercises[index], { weight: event.target.value });
```

this applies to the reps and completed status too